### PR TITLE
docs(idempotency): API docs for idempotency

### PR DESCRIPTION
**Pull request by am29d**
## Description of your changes

* bump typedoc deps
* add typedoc config per package with a base configuration in the root folder, this helps us to control the `entryPoints` for each package
* add api docs for idempotency
* changed idempotency README to specifics of that package and additional information to the root README. 

### Related issues, RFCs

**Issue number:** #1104 

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-typescript/#tenets)
- [ ] I have performed a *self-review* of my own code
- [ ] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [ ] My changes generate *no new warnings*
- [ ] I have *added tests* that prove my change is effective and works
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

***Is it a breaking change?:*** NO

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: tsconfig.json
Comment (dreamorosi) [inline tsconfig.json:1]: Why was this needed here?

I don't disagree that we should also centralize the `tsconfig` and apply the same pattern of extending it (#617), just trying to understand if there's a dependency between the changes made in this PR and this file being at the root.

File: typedoc.base.json
Comment (dreamorosi) [inline typedoc.base.json:1]: What's the difference between `typedoc.json` and `typedoc.base.json` and do we need both?

File: packages/idempotency/src/persistence/index.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
 export * from './BasePersistenceLayer';
 export * from './BasePersistenceLayerInterface';
 export * from './IdempotencyRecord';
+export * from './DynamoDBPersistenceLayer';
```
Comment (dreamorosi) [inline packages/idempotency/src/persistence/index.ts:4]: I left this out and exposed it via a separate import on purpose, and I feel strongly (but are willing to discuss) about keeping it that way.

At the moment we only have one persistence layer, so having it exported here is fine.

If (or actually when) we add more persistence layers (i.e. ElastiCache or others), this file will become:

```ts
// packages/idempotency/src/persistence/index.ts
export * from './DynamoDBPersistenceLayer';
export * from './OtherPersistenceLayer';
```

However **customers using one won't need the other**.

By having it like this, whenever customers do this: `import { DynamoDBPersistenceLayer } from '@aws-lambda-powertools/idempotency/persistence';` or this: `import { OtherPersistenceLayer } from '@aws-lambda-powertools/idempotency/persistence';` both are imported.

This a few undesirable effects:

- If any of the persistence layers have unique dependencies, compilers like `tsc` will require them to be installed in the project, even if you are not explicitly using the code (because you are importing it) - this has happened in #1068, resulted in #1225, and is something that we [want to change in v2](https://github.com/awslabs/aws-lambda-powertools-typescript/discussions/1269#discussioncomment-4891876) for Logger, Tracer, and Metrics.
- Some bundlers like `esbuild` cannot always tree shake the unused module, because it's always being imported
- If the initialization of the module is expensive (which is **not** the case with DynamoDB), then customers would be paying this price for both persistence layers.

By having them separate instead, customers have to explicitly opt-in and import whichever persistence layer they are actually using, i.e. `import { DynamoDBPersistenceLayer } from '@aws-lambda-powertools/idempotency/dynamodb';` or `import { OtherPersistenceLayer } from '@aws-lambda-powertools/idempotency/other';`. This way the imports are isolated from each other and: 1/ dependencies for each one can be installed selectively, 2/ tree-shaking and bundling can be done effectively.

This can be achieved with customizing the exports ([see here](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/packages/idempotency/package.json#L44-L47)), and is also why in Parameters we don't have an `index.ts` file that exports all the different providers.

File: packages/idempotency/src/persistence/IdempotencyRecord.ts [hunk 2]
```diff
@@ -1,15 +1,31 @@
...
   public inProgressExpiryTimestamp?: number;
+  /**
+   * The hash of the payload of the request, used for comparing requests.
+   */
   public payloadHash?: string;
+  /**
+   * The response data of the request, this will be returned if the payload hash matches.
+   */
   public responseData?: Record<string, unknown>;
   private status: IdempotencyRecordStatus;
```
Comment (dreamorosi) [inline packages/idempotency/src/persistence/IdempotencyRecord.ts:30]: Can we add a comment here like the others & maybe link to `IdempotencyRecordStatus`?

File: packages/idempotency/src/index.ts [hunk 3]
```diff
@@ -2,3 +2,5 @@ export * from './Exceptions';
 export * from './IdempotencyConfig';
 export * from './idempotentDecorator';
 export * from './makeFunctionIdempotent';
+export * from './middleware';
```
Comment (dreamorosi) [inline packages/idempotency/src/index.ts:5]: See comment I left under `packages/idempotency/src/persistence/index.ts` about imports/exports - TL;DR; the omission was intentional and I believe should stay like that.

File: packages/idempotency/README.md [hunk 4]
```diff
@@ -1,71 +1,131 @@
...
 
-The [Serverless TypeScript Demo](https://github.com/aws-samples/serverless-typescript-demo) shows how to use Powertools for AWS Lambda (TypeScript).  
-You can find instructions on how to deploy and load test this application in the [repository](https://github.com/aws-samples/serverless-typescript-demo).
+const processIdempotently = makeFunctionIdempotent(proccessingFunction, {
+  persistenceStore: dynamoDBPersistenceLayer,
+  dataKeywordArgument: "transactionId"
+});
+
+export const handler = async (
+  _event: Event,
```
Comment (dreamorosi) [inline packages/idempotency/README.md:121]: Ideally these examples should be copy-paste-able. If we I do that I don't have imports for `Event` & `Context`.

We can either add the imports from `aws-lambda`, or use `unknown` like done in the rest of the docs/examples.

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Hey Alex, thanks for the great work - I appreciate that you took the time to improve the API docs for the whole project and standardized the configs!

I have left a few comments, mostly minor ones, except one related to the imports and exports. See more details below - happy to discuss it further later if needed.

File: typedoc.base.json
Comment (am29d) [reply typedoc.base.json:1]: `typedoc.base.json` is the common configuration for individual packages. We need to the separation because otherwise typedoc will have a circular dependency if you run it on the root level: `typedoc.json` -> `package/logger/typedoc.json-> `typedoc.json`. The error sounds something like this:

> [error] Project at ./packages/logger has entryPointStrategy set to packages, but nested packages are not supported.

File: packages/idempotency/src/persistence/index.ts (hunk 1 above)
Comment (am29d) [reply packages/idempotency/src/persistence/index.ts:4]: Hmm I see, I was wondering about the inconsistency of the exports. I agree that we should be mindful of the exports. 

I added it to collect it for the typedoc output, since the entrypoint is `index.ts`. I will revert the exports to the previous state.

File: packages/idempotency/README.md (hunk 4 above)
Comment (am29d) [reply packages/idempotency/README.md:121]: you are right, I always complained about missing import in any other docs, now I did the same 😅

File: typedoc.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+  "entryPoints": ["packages/*"],
+  "entryPointStrategy": "packages",
+  "name": "AWS Lambda Powertools Typescript API Reference",
```
Comment (dreamorosi) [inline typedoc.json:4]:
```suggestion
  "name": "Powertools for AWS Lambda (Typescript) API Reference",
```

**Review by am29d — COMMENTED** [review]
good catch

**Review by dreamorosi — APPROVED** [review]
Good work with this PR, the new docs are looking nice.

I think we should include the `BaseProvider` class for Parameters (so that customers can extend it and build their own provider) and check that `DynamoDBPersistenceLayer` is included in the exports - but we'll do it in separate PRs.
